src/core/string.h

```diff
@@ -0,0 +1,231 @@
+// Copyright (c) 2014-2015 Dropbox, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef PYSTON_CORE_STRING_H
+#define PYSTON_CORE_STRING_H
+
+#include <stdexcept>
+
+namespace pyston {
+
+class String {
+public:
+    explicit String(const std::string& s, char* _ptr = NULL) : ptr(_ptr), own(_ptr == NULL) {
```

Comment:
I found and fixed the cases where we were creating temporaries (operator+, operator<<), and switched those instances to using llvm::Twines and llvm::raw_string_ostream.  The latter lets us use llvm::StringRef directly, and the former delays the concatenation until the place where the value is used (in the new `boxStringTwine` method).  Now there are no cases where pyston::String owns the buffer, so it shrinks by a bool, the ctors get simpler, and there's no dtor.